desc/objectives/_stability.py

```diff
@@ -367,3 +369,213 @@ def compute(self, params, constants=None):
+        self._len_keys = ["a"]
+        len_profiles = get_profiles(self._len_keys, obj=eq, grid=len_grid)
+        len_transforms = get_transforms(self._len_keys, obj=eq, grid=len_grid)
+
+        # make a set of nodes along a single fieldline
+        zeta = np.linspace(-self.zetamax, self.zetamax, self.nzeta)
+        rho, alpha, zeta = np.reshape(
+            np.broadcast_arrays(self.rho, self.alpha, zeta), (3, -1)
+        )
+        fieldline_nodes = np.array([rho, alpha, zeta]).T
+
+        self._dim_f = 1
+        self._data_keys = ["ideal_ball_gamma2"]
+
+        self._args = get_params(
+            self._iota_keys + self._len_keys + self._data_keys,
+            obj="desc.equilibrium.equilibrium.Equilibrium",
+            has_axis=False,
+        )
+
+        self._constants = {
+            "iota_transforms": iota_transforms,
+            "iota_profiles": iota_profiles,
+            "len_transforms": len_transforms,
+            "len_profiles": len_profiles,
+            "fieldline_nodes": fieldline_nodes,
+            "quad_weights": 1.0,
+        }
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, params, constants=None):
+        """
+        Compute the ballooning stability growth rate.
+
+        Parameters
+        ----------
+        params : dict
+            Dictionary of equilibrium degrees of freedom, eg Equilibrium.params_dict
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc. Defaults to
+            self.constants
+
+        Returns
+        -------
+        gamma : ndarray
+            ideal ballooning growth rate.
+
+        """
+        from desc.equilibrium.coords import map_coordinates
+
+        eq = self.things[0]
+
+        if constants is None:
+            constants = self.constants
+        # we first compute iota on a uniform grid to get correct averaging etc.
+        iota_data = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self._iota_keys,
+            params=params,
+            transforms=constants["iota_transforms"],
+            profiles=constants["iota_profiles"],
+        )
+
+        len_data = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self._len_keys,
+            params=params,
+            transforms=constants["len_transforms"],
+            profiles=constants["len_profiles"],
+        )
+
+        # Now we compute theta_DESC for given theta_PEST
+        iota = iota_data["iota"][0]
+        rho, alpha, zeta = constants["fieldline_nodes"].T
+        theta_PEST = alpha + iota * zeta
+        nodes = jnp.array([rho, theta_PEST, zeta]).T
+
+        # Rootfinding theta for a given theta_PEST
+        desc_coords = map_coordinates(eq, nodes, inbasis=("rho", "theta_PEST", "zeta"))
+
+        diff_arr = jnp.diff(desc_coords[:, :], axis=0)
+        nodes_spacing = jnp.concatenate([diff_arr, diff_arr[-1:, :]], axis=0)
```

Comment:
@rahulgaur104 in [f2e18dd](https://github.com/PlasmaControl/DESC/pull/1170/commits/f2e18dd8a86318efeced3ac77966845e88e96fcd) you removed `iota_r` from the list of quantities that are calculated on a uniform tensor-product grid in DESC coordinates. The relevant lines are [here](https://github.com/PlasmaControl/DESC/pull/1170/commits/f2e18dd8a86318efeced3ac77966845e88e96fcd#r1717340379) and [here](https://github.com/PlasmaControl/DESC/pull/1170#discussion_r1717337004). 

You should revert this and remove nodes_spacing.

Here is longer explanation:

The changes aren't good because currently `iota_r` will be computed on your non-uniform straight field line grid `sfl_grid`, which will be signficantly less accurate do to worse convergence properties of the quadrature. Since `iota_r` is calculated poorly, so will other quantities that your stability quantity `ideal_balloning_gamma_2` depends on such as components of the B field.

To notify you of this, the code raised a warning saying that `sfl_grid` doesn't have the proper quadrature weights to perform the integration, that is the  `_spacing` attribute is missing. The patch of adding quadrature weights to the grid will suppress this warning, but it's a bad choice to do the quadrature on a grid with non-uniform spacing.

Reply:
I have created a new issue to update the warning to suggest  users toward computing things on a uniform grid first when they see this warning, rather than trying to construct quadrature weights on a non-uniform grid.

Reply:
Ok, I have added your suggestion and the test runs now. I will push the changes after the dev meeting.

Reply:
Done.